designsafe/apps/workspace/api/views.py

```diff
@@ -113,18 +117,33 @@ def _get_app(app_id, app_version, user):
     return data
 
 
-def test_system_needs_keys(tapis, system_id):
+def test_system_needs_keys(tapis, username, system_id):
     """Tests a Tapis system by making a file listing call.
 
-    returns: SystemDef
-    """
+    If the system is TMS_KEYS-based, it attempts to create credentials before listing files.
+
+    Args:
+        tapis (Tapis): An instance of the Tapis client.
+        username (str)): The user to create credentials.
+        system_id (str): The ID of the Tapis system to test.
 
+    Returns:
+        SystemDef: The system definition if an error occurs.
+    """
+    system_def = tapis.systems.getSystem(systemId=system_id)
     try:
         tapis.files.listFiles(systemId=system_id, path="/")
+        return None
     except (InternalServerError, UnauthorizedError):
-        system_def = tapis.systems.getSystem(systemId=system_id)
+        # Check if the system uses TMS_KEYS and create credentials if necessary
+        if system_def.get("defaultAuthnMethod") == 'TMS_KEYS':
+            create_system_credentials(tapis, username, system_id, createTmsKeys=True)
```

Comment:
I think we need to wrap the `try/except` around `create_system_credentials` followed by `listFiles`, since by default the `createSystemCredentials` call will actually throw an exception if the credentials don't work